opencv 3.1.0: add patch fixing python use of FlannBasedMatcher.add

risicle opened the pull request:
###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Patch is from https://github.com/opencv/opencv/pull/6009, corresponding to
opencv commit 05cfe28612fd8dc8fb0ccb876df945c7b435dd26. Upstream doesn't seem
particularly enthusiastic about a 3.1.x point release, so who knows when
this fix would otherwise see the light of day.

mdaiter commented:
@risicle, was there no existing commit straight from the OpenCV repository that would fix this issue? For example, the previous fixes for CUDA 8 were able to be added just by referencing two commits in the OpenCV repo.
Just don't want to have patch files lying around if we can use a straight commit from the OpenCV repo instead: it's a bit easier to maintain.

risicle commented:
Has it *really* been decided that that's a better way of doing it? I've been hamstrung in the past by someone choosing a dumb source for a patch on an unreliable server (and github's been known to have its outages...) and been unable to build my system.

risicle commented:
I'll do it if that's *really* thought to be a better way, though.

(^ "no space left on device" travis failure)

mdaiter commented:
@risicle I just don't want patch files lying around in the directory, as I don't know the intricacies of the files or the effect they have on the source code.
If it's commit 05cfe28612fd8dc8fb0ccb876df945c7b435dd26 from the OpenCV repo , we should add that to the patches array.
Also, this would still not build if GitHub went down, as we pull two patches from GitHub as well as this patch.

risicle committed:
opencv 3.1.0: add patch fixing python use of FlannBasedMatcher.add

Patch is from https://github.com/opencv/opencv/pull/6009. Upstream doesn't
seem particularly enthusiastic about a 3.1.x point release, so who knows
when this fix would otherwise see the light of day.

risicle commented:
Cool. Guess I wouldn't be making it any worse. New version pushed.

risicle commented:
(again, "no space left on device" travis failure)

mdaiter commented:
@risicle , shouldn't be a problem. Happened to me as well when I was updating OpenCV 3 with my CUDA-related patches (https://github.com/NixOS/nixpkgs/pull/20631) .

cc @viric for approval and merge.

viric commented:
What application problem is this currently fixing? We would not just pick all upstream changes into nixpkgs.

risicle commented:
Well, it's a bug that stops `FlannBasedMatcher` working from the python bindings and upstream hasn't made a release for quite a while, so this is the only way of having a working python `FlannBasedMatcher` in nix. And as for applications, well, there aren't any applications in nix depending on opencv, so I guess that would make opencv the application itself, no?

mdaiter commented:
@risicle , could you make a conditional to check if Python support is selected, and if so, then you apply this patch?
I should also probably do the same with my CUDA patches.
This is to ensure, as @viric stated, that we don't pull in patches that the end user doesn't need. We shouldn't have to download/apply patches that won't affect the experience or the quality of the final product for the end-user.

@viric , comments on finalising this merge?

risicle commented:
@mdaiter good idea, will do.

risicle commented:
@abbradar ? @viric ? thoughts on merging?

grahamc commented:
Builds ok
